solr/core/src/java/org/apache/solr/schema/WrappedFieldValueSource.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.schema;
+
+import java.io.IOException;
+import java.util.Map;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.queries.function.FunctionValues;
+import org.apache.lucene.queries.function.ValueSource;
+import org.apache.lucene.search.DoubleValuesSource;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.LongValuesSource;
+import org.apache.lucene.search.SortField;
+
+/**
+ * Wraps a ValueSource that derives values directly from a particular SchemaField. This allows a
+ * relatively clean way to set missingValue on ValueSource.getSortField, and also clearly marks
+ * SortFields that are directly associated with a SchemaField (e.g., for purposes of
+ * marshaling/unmarshaling sort values).
+ */
+public class WrappedFieldValueSource extends ValueSource {
+
+  private final SchemaField f;
+  private final ValueSource backing;
+
+  public WrappedFieldValueSource(SchemaField f, ValueSource backing) {
+    this.f = f;
+    this.backing = backing;
+  }
+
+  public ValueSource unwrap() {
+    return backing;
+  }
+
+  public SchemaField getSchemaField() {
```

Comment:
I notice this wrapper just carries a schemaField. Why we can't pull this method up to ValueSource, and stub it `return null` by default? This allows to avoid those `instanceof`s. They are basically ok, but, you know, lacking of elegance. Or perhaps: 
`Optional<SchemaField> ValueSource::getSchemaField()` ?

Reply:
Yes that makes sense I think. I'll see what I can do and report back.